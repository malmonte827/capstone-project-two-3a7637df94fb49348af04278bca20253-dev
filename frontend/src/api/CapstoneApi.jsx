import axios from "axios"

const BASE_URL = process.env.React_APP_BASE_URL || "http://localhost:3001"

/** APi class
 * 
 * Staic class to get/send to the API
 */

class CapstoneApi{
    // token for interacting with the API
    static token

    static async request(endpoint, data={}, method="get"){
        console.debug(`API Call:`, endpoint, data, method)

        const url = `${BASE_URL}/${endpoint}`
        const headers = {Authorization: `Bearer ${CapstoneApi.token}`}
        const params = (method === "get") ? data : {}

        try{
            return (await axios({url, method, data, params,headers})).data
        }catch(err){
            console.error(`API Error:`, err.response)
            let message = err.response.data.error.message
            throw Array.isArray(message) ? message : [message]
        }
    }

    /** API ROUTES */

    /** Get the current user */

    static async getCurrentUser(username){
        let res = await this.request(`/users/${username}`)
        return res.user
    }

/** Signup User */
    static async signup(data){
        let res = await this.request(`/auth/register`, data, "post")
        return res.token
    }

    /** Login User */
    static async login(data){
        const res = await this.request(`/auth/token`, data, "post")
        return res.token
    }
}

export default CapstoneApi