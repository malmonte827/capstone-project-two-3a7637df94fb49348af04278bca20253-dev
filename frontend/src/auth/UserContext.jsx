import React from "react"

/** COntext provides current user object throughout app */

const UserContext = React.createContext()

export default UserContext