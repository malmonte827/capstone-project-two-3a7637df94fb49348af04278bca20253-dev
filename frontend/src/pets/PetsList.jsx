import React, { useState, useEffect, useContext } from "react";
import CapstoneApi from "../api/CapstoneApi";
import PetCard from "./PetCard";
import LoadingSpinner from "../common/LoadingSpinner";
import UserContext from "../auth/UserContext";
import PetForm from "./PetForm"
import { Link } from "react-router-dom";

/**
 * PetList Component
 *
 * Displays a list  of all the users pets
 * Provides buttons to edit or delete the pet.
 *
 * Props:
 * - pet (object): The pet object containing pet details (pet_id, name, age, species, hunger).
 * - onDelete (function): A callback function to re-render the list after a pet is deleted.
 *
 * Context:
 * - currentUser (object): The logged-in user's information from UserContext.
 */

function PetList() {
    const { currentUser } = useContext(UserContext);
    const [pets, setPets] = useState(null);
    const [error, setError] = useState(null);

     // Fetches the user's pets on mount.
    useEffect(() => {
        getPets()
    }, [])

 /**
     * Retrieves pets from the API.
     * Updates state with the list of pets or handles errors.
     */
        async function getPets() {
            try{
        let pets = await CapstoneApi.getPets(currentUser.username);
        setPets(pets);
        setError(null)
            }catch(err){
                console.error("Error getting pets", err)
                setError(err)
            }
    }

    if (!pets) return <LoadingSpinner />;

    return (
        <div className="PetsList col-md-8 offset-md-2">
            {pets.length ? (
                <div className="PetsList-list">
                    {pets.map((p) => (
                        <PetCard
                            pet={p}
                            key={p.pet_id}
                            onDelete={getPets}
                        />
                    ))}
                </div>
            ) : (
                <p className="lead">Sorry, no pets were found!</p>
                
    
            )}
        <div>
            <button><Link to="/pets/add">Click to add pet</Link></button>
            
        </div>
        </div>
    );
}
export default PetList;
