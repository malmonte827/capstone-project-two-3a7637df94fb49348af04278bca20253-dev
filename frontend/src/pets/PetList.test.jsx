import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import "@testing-library/jest-dom";
import { expect, vi } from "vitest";
import PetsList from "./PetsList";
import UserContext from "../auth/UserContext";
import CapstoneApi from "../api/CapstoneApi";

vi.mock("../api/CapstoneApi");

describe("<PetList />", () => {
    const currentUser = { username: "testuser" };
    const pets = [
        { pet_id: 1, name: "Buddy", age: 5, species: "Dog", hunger: 3 },
        { pet_id: 2, name: "Mittens", age: 3, species: "Cat", hunger: 5 },
    ];

    it("renders without crashing", () => {
        render(
            <MemoryRouter>
                <UserContext.Provider value={{ currentUser }}>
                    <PetsList />
                </UserContext.Provider>
            </MemoryRouter>
        );
    });

    it("matches snapshot", async () => {
        CapstoneApi.getPets.mockResolvedValueOnce(pets);

        const { asFragment } = render(
            <MemoryRouter>
                <UserContext.Provider value={{ currentUser }}>
                    <PetsList />
                </UserContext.Provider>
            </MemoryRouter>
        );
        await waitFor(() =>
            expect(screen.getAllByText("Buddy").length).toBeGreaterThan(0)
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders pet cards when pets are available", async () => {
        CapstoneApi.getPets.mockResolvedValueOnce(pets);

        render(
            <MemoryRouter>
                <UserContext.Provider value={{ currentUser }}>
                    <PetsList />
                </UserContext.Provider>
            </MemoryRouter>
        );

        await waitFor(() => {
            expect(screen.getAllByText("Buddy").length).toBeGreaterThan(0);
            expect(screen.getAllByText("Mittens").length).toBeGreaterThan(0);
        });
    });

    it("shows message when no pets are found", async () => {
        CapstoneApi.getPets.mockResolvedValueOnce([]);

        render(
            <MemoryRouter>
                <UserContext.Provider value={{ currentUser }}>
                    <PetsList />
                </UserContext.Provider>
            </MemoryRouter>
        );

        await waitFor(() => {
            expect(
                screen.getByText("Sorry, no pets were found!")
            ).toBeInTheDocument();
        });
    });
});
