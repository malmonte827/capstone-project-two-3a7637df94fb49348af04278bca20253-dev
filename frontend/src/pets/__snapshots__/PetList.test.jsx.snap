// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<PetList /> > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="PetsList col-md-8 offset-md-2"
  >
    <div
      class="PetsList-list"
    >
      <div
        class="PetCard card"
      >
        <div
          class="container"
        >
          <div
            class="card"
          >
            <div
              class="card-head"
            >
              <div
                class="animal-detail"
              >
                <h2>
                  Buddy
                </h2>
                <img
                  alt="Pet Image"
                  class="animal-img"
                  src="../image/download.jpg"
                />
              </div>
            </div>
            <div
              class="card-body"
            >
              <div
                class="animal-title"
              >
                <h1>
                  Buddy 
                  <span
                    class="badge"
                  >
                     Dog 
                  </span>
                </h1>
              </div>
              <div
                class="animal-info"
              >
                <h4>
                  Info
                </h4>
                <ul>
                  <li>
                    <span
                      class="labels"
                    >
                      Hunger Level:
                    </span>
                     3
                  </li>
                  <li>
                    <span
                      class="labels"
                    >
                      Age:
                    </span>
                     5 years.
                  </li>
                </ul>
              </div>
              <div
                class="animal-description"
              >
                <h4>
                  Description
                </h4>
                <p />
              </div>
              <div
                class="mt-4"
              >
                <a
                  class="btn btn-primary"
                  data-discover="true"
                  href="/pets/edit"
                >
                  Edit
                </a>
                <button
                  class="btn btn-danger"
                >
                  Delete
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="PetCard card"
      >
        <div
          class="container"
        >
          <div
            class="card"
          >
            <div
              class="card-head"
            >
              <div
                class="animal-detail"
              >
                <h2>
                  Mittens
                </h2>
                <img
                  alt="Pet Image"
                  class="animal-img"
                  src="../image/download.jpg"
                />
              </div>
            </div>
            <div
              class="card-body"
            >
              <div
                class="animal-title"
              >
                <h1>
                  Mittens 
                  <span
                    class="badge"
                  >
                     Cat 
                  </span>
                </h1>
              </div>
              <div
                class="animal-info"
              >
                <h4>
                  Info
                </h4>
                <ul>
                  <li>
                    <span
                      class="labels"
                    >
                      Hunger Level:
                    </span>
                     5
                  </li>
                  <li>
                    <span
                      class="labels"
                    >
                      Age:
                    </span>
                     3 years.
                  </li>
                </ul>
              </div>
              <div
                class="animal-description"
              >
                <h4>
                  Description
                </h4>
                <p />
              </div>
              <div
                class="mt-4"
              >
                <a
                  class="btn btn-primary"
                  data-discover="true"
                  href="/pets/edit"
                >
                  Edit
                </a>
                <button
                  class="btn btn-danger"
                >
                  Delete
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div>
      <a
        class="btn btn-primary"
        data-discover="true"
        href="/pets/add"
      >
        Click to add pet
      </a>
    </div>
  </div>
</DocumentFragment>
`;
